keps/sig-network/20190613-kube-proxy-component-config.md

```diff
@@ -0,0 +1,210 @@
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ...
+modeB: ...
+modeC: ...
+```
+
+### Group platform specific fields together
+
+Platform specific fields need to be grouped together by platform name. This can be done by introducing structs to house them - one per platform (one for Linux, one for Windows, etc.). Then these structs will have an instance in the KubeProxyConfiguration.
+That can allow for users to specify settings for more than one platform per a YAML document, thus allowing for deduplication of platform independent shared fields. This can reduce the possibility of an error for users.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ... # modeA is platform independent
+platformX:
+  modeB: ...
+platformY:
+  modeC: ...
+```
+
+### Split instance specific and shared settings
+
+As a last step, instance specific and shared settings should be split into a couple of different YAML documents. This will allow for such settings to be specified in one or two YAML files, depending on user preferences.
+Preliminary reviews of the impacted settings show, that only settings, that are not platform or mode specific are instance specific. Therefore, the existing platform-mode specific structures are to be preserved in the shared settings.
+However, should the need arise, the platform-mode specific structures should be used in the instance specific structure as well.
+
+#### Example
+
+```yaml
+kind: KubeProxyInstanceConfiguration
+setting1: ...
+setting2: ...
+...
+modeA: ... # modeA is platform independent
+platformX:
+  modeB: ...
+platformY:
+  modeC: ... # no modeD, as modeD does not have instance specific settings
+---
+kind: KubeProxySharedConfiguration
+sharedSetting1: ...
+sharedSetting2: ...
+...
+modeA: ... # modeA is platform independent
+platformX:
+  modeB: ...
+platformY:
+  modeD: ... # no modeC, as modeC does not have shared settings
+```
+
+### Risks and Mitigations
+
+So far, the following risks have been identified:
+- Deviation of the implementation guidelines and bad planning may have the undesired effect of producing bad alpha versions.
+- Bad alpha versions will need good alpha versions to fix them. This will create too many iterations over the API and users will get confused.
+- New and redesigned kube-proxy API versions may cause confusion among users who are used to the v1alpha1 relatively flat, single document design. In particular, multiple YAML documents and structured (as opposed to flat) objects can create confusion as to what option is placed where.
+
+The mitigations to those risks:
+- Strict following of the proposals in this document and planning ahead for a release and config cycle.
+- Support reading from the last couple of API versions released. When the beta version is released, support the last alpha version for one or two release cycles after that.
```

Comment:
That was bad wording on my part. It should not be renamed, but instead, we should have a copy, that says "v1beta1".

If the alpha configs are short lived they won't see too much use. However, I would expect that in the case of kube-proxy, people are going to notice the deletion of v1alpha1 and this can cause a few problems if we force them to use a new config version without prior notice.